{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "089cf773-5a8e-47e8-8fb8-04054b195cb2",
   "metadata": {},
   "source": [
    "# Install structsense and makesure GROBID is running. Use the command below.\n",
    "\n",
    "```shell\n",
    "docker pull lfoppiano/grobid:0.8.0\n",
    "docker run --init -p 8070:8070 -e JAVA_OPTS=\"-XX:+UseZGC\" lfoppiano/grobid:0.8.0\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a637454f-04d0-44cc-a1c4-7235bee34310",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install structsense --quiet "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c64a6973-fbbc-479f-bf5c-85725f94541f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: structsense-cli [OPTIONS] COMMAND [ARGS]...\n",
      "\n",
      "  CLI commands for the Structsense Framework application\n",
      "\n",
      "Options:\n",
      "  --help  Show this message and exit.\n",
      "\n",
      "Commands:\n",
      "  extract  Extract the terms along with sentence.\n"
     ]
    }
   ],
   "source": [
    "!structsense-cli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6801df16-eab8-4c36-91fa-3a69010f4e7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: structsense-cli extract [OPTIONS]\n",
      "\n",
      "  Extract the terms along with sentence.\n",
      "\n",
      "Options:\n",
      "  --agentconfig TEXT      Path to the agent configuration in YAML file format\n",
      "                          or dictionary  [required]\n",
      "  --taskconfig TEXT       Path to the agent task configuration in YAML format\n",
      "                          or or dictionary  [required]\n",
      "  --embedderconfig TEXT   Path to the embedding configuration in YAML format\n",
      "                          or or dictionary  [required]\n",
      "  --flowconfig TEXT       Path to the flow configuration in YAML format or or\n",
      "                          dictionary. The flow configuration describes the\n",
      "                          flow of the agent.  [required]\n",
      "  --knowledgeconfig TEXT  Path to the configuration in YAML format or or\n",
      "                          dictionary that specify the search knowledge search\n",
      "                          key.\n",
      "  --source TEXT           The sourceâ€”whether a file (text or PDF), a folder,\n",
      "                          or a text string.  [required]\n",
      "  --help                  Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "!structsense-cli extract --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d5af4c7-7cba-4bd3-a375-460acfd07a96",
   "metadata": {},
   "source": [
    "# Set Environment Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8490e7fb-49e4-4a5e-995b-db14a7696970",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: ENABLE_WEIGHTSANDBIAS=false\n",
      "env: ENABLE_MLFLOW=false\n",
      "env: ENABLE_KG_SOURCE=false\n",
      "env: ONTOLOGY_DATABASE=Ontology_database_agent_test #note this is my local vectordatabase name, we need this because ENABLE_KG_SOURCE=true\n",
      "env: WEAVIATE_API_KEY=\"XFG!NQmUVEC&8\" #note this is my local vectordatabase key\n",
      "env: OLLAMA_API_ENDPOINT=http://host.docker.internal:11434 #note this is my local vectordatabase name\n",
      "env: OLLAMA_MODEL=nomic-embed-text #note this is my local vectordatabase name\n"
     ]
    }
   ],
   "source": [
    "%env ENABLE_WEIGHTSANDBIAS=false\n",
    "%env ENABLE_MLFLOW=false\n",
    "%env ENABLE_KG_SOURCE=false\n",
    "%env ONTOLOGY_DATABASE=Ontology_database_agent_test #note this is my local vectordatabase name, we need this because ENABLE_KG_SOURCE=true\n",
    "%env WEAVIATE_API_KEY=\"XG!mUEC&io\" #note this is my local vectordatabase key\n",
    "%env OLLAMA_API_ENDPOINT=http://host.docker.internal:11434 #note this is my local vectordatabase name\n",
    "%env OLLAMA_MODEL=nomic-embed-text #note this is my local vectordatabase name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee709161-3acd-4536-b9b9-2833c92379ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mcrew_memory\u001b[m\u001b[m           \u001b[34mner_config\u001b[m\u001b[m            structsense-cli.ipynb\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4effa86e-bba4-4b5d-9175-a96acb1153ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "embedding.yaml                 ner_task.yaml\n",
      "flow_ner.yaml                  search_ontology_knowledge.yaml\n",
      "ner_agent.yaml\n"
     ]
    }
   ],
   "source": [
    "!ls ner_config/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "54a9254d-02e3-4dde-9de0-bb0845441acd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "agents:\n",
      "  - id: extractor_agent\n",
      "    output_variable: extracted_info\n",
      "    role: >\n",
      "      Neuroscience Named Entity Recognition (NER) Extractor Agent\n",
      "    goal: >\n",
      "      Perform Named Entity Recognition (NER) on neuroscience {literature} and return structured JSON output.\n",
      "    backstory: >\n",
      "      You are an AI assistant specialized in processing neuroscience and who do not hallucinate. \n",
      "      Your expertise includes recognizing and categorizing named entities such as anatomical regions, experimental conditions, and cell types. \n",
      "      Your responses strictly adhere to JSON format, ensuring accurate and structured data extraction for downstream applications.\n",
      "    llm:\n",
      "      model: ollama/deepseek-r1:14b\n",
      "      base_url: http://localhost:11434\n",
      "      seed: 53\n",
      "\n",
      "  - id: alignment_agent\n",
      "    output_variable: aligned_structured_terms\n",
      "    role: >\n",
      "      Neuroscience Named Entity Recognition (NER) Concept Alignment Agent\n",
      "    goal: >\n",
      "      Perform concept alignment to the extracted Named Entity Recognition (NER) by extractor_agent {extracted_info} and return structured JSON output.\n",
      "    backstory: >\n",
      "      You are an AI assistant specialized in processing neuroscience concept alignment with structured models, i.e., ontologies or schemas and who do not hallucinate. \n",
      "      Your expertise includes recognizing and categorizing extracted named entities such as anatomical regions, experimental conditions, and cell types and aligning the recognized named entities such as cell types with corresponding ontological terms. \n",
      "      Your responses strictly adhere to JSON format, ensuring accurate and structured data extraction for downstream applications.\n",
      "    llm:\n",
      "      model: ollama/deepseek-r1:14b\n",
      "      base_url: http://localhost:11434\n",
      "      seed: 53\n",
      "\n",
      "  - id: judge_agent\n",
      "    output_variable: aligned_judged_terms\n",
      "    role: >\n",
      "      Neuroscience Named Entity Recognition (NER) Judge Agent\n",
      "    goal: >\n",
      "      Evaluate the {aligned_structured_terms} based on predefined criteria and generate a structured JSON output reflecting the assessment results.\n",
      "    backstory: >\n",
      "      You are an AI assistant with expert knowledge in neuroscience and structured models, i.e., ontologies or schemas, and someone who does not hallucinate.  \n",
      "      Your task is to evaluate the {aligned_structured_terms} based on the accuracy and quality of the alignment. \n",
      "      Assign the score between 0-1 with 1 being the highest score of your evaluation.\n",
      "      Your responses strictly adhere to JSON format, ensuring accurate and structured data extraction for downstream applications.\n",
      "    llm:\n",
      "      model: ollama/deepseek-r1:14b\n",
      "      base_url: http://localhost:11434\n",
      "      seed: 53\n"
     ]
    }
   ],
   "source": [
    "! cat ner_config/ner_agent.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93626f4f-7cc4-4748-8467-1b3cb4a9a067",
   "metadata": {},
   "source": [
    "# Extract from a PDF file (with knowledge source)\n",
    "\n",
    "Input: https://www.biorxiv.org/content/10.1101/2025.03.06.641914v1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3905a1bf-4bbb-47a9-a359-260d5c7f1688",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-04-02 15:33:41,800 - structsense.cli - INFO - Processing source: test.pdf with agent config: ner_config/ner_agent.yaml, task config: ner_config/ner_task.yaml, embedderconfig: ner_config/embedding.yaml, flowconfig: ner_config/flow_ner.yaml, knowledgeconfig: ner_config/search_ontology_knowledge.yaml\n",
      "Processing source: test.pdf with agent config: ner_config/ner_agent.yaml, task config: ner_config/ner_task.yaml, embedderconfig: ner_config/embedding.yaml knowledgeconfig: ner_config/search_ontology_knowledge.yaml flowconfig: ner_config/flow_ner.yaml\n",
      "2025-04-02 15:33:41,801 - utils.utils - INFO - Trying paths: ['test.pdf', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/test.pdf', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/test.pdf', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/test.pdf']\n",
      "2025-04-02 15:33:41,801 - utils.utils - INFO - Using path: test.pdf\n",
      "2025-04-02 15:33:41,801 - utils.utils - INFO - Processing single file: test.pdf\n",
      "2025-04-02 15:33:46,033 - utils.utils - INFO - Successfully extracted 17 sections\n",
      "\u001b[34mâ•­â”€\u001b[0m\u001b[34mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\u001b[0m\u001b[34m Flow Execution \u001b[0m\u001b[34mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\u001b[0m\u001b[34mâ”€â•®\u001b[0m\n",
      "\u001b[34mâ”‚\u001b[0m                                                                              \u001b[34mâ”‚\u001b[0m\n",
      "\u001b[34mâ”‚\u001b[0m  \u001b[1;34mStarting Flow Execution\u001b[0m                                                     \u001b[34mâ”‚\u001b[0m\n",
      "\u001b[34mâ”‚\u001b[0m  \u001b[37mName: \u001b[0m\u001b[34mStructSenseFlow\u001b[0m                                                       \u001b[34mâ”‚\u001b[0m\n",
      "\u001b[34mâ”‚\u001b[0m  \u001b[37mID: \u001b[0m\u001b[34m5931eff8-4423-41a8-a6e8-05e70ed871ee\u001b[0m                                    \u001b[34mâ”‚\u001b[0m\n",
      "\u001b[34mâ”‚\u001b[0m                                                                              \u001b[34mâ”‚\u001b[0m\n",
      "\u001b[34mâ”‚\u001b[0m                                                                              \u001b[34mâ”‚\u001b[0m\n",
      "\u001b[34mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯\u001b[0m\n",
      "\n",
      "2025-04-02 15:33:46,038 - utils.utils - INFO - Trying config paths: ['ner_config/ner_agent.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/ner_agent.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/ner_agent.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/ner_agent.yaml']\n",
      "2025-04-02 15:33:46,048 - utils.utils - INFO - file processing - <_io.TextIOWrapper name='ner_config/ner_agent.yaml' mode='r' encoding='utf-8'>, type: agent\n",
      "2025-04-02 15:33:46,049 - utils.utils - INFO - Trying config paths: ['ner_config/ner_task.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/ner_task.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/ner_task.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/ner_task.yaml']\n",
      "2025-04-02 15:33:46,066 - utils.utils - INFO - file processing - <_io.TextIOWrapper name='ner_config/ner_task.yaml' mode='r' encoding='utf-8'>, type: task\n",
      "2025-04-02 15:33:46,066 - utils.utils - INFO - Trying config paths: ['ner_config/embedding.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/embedding.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/embedding.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/embedding.yaml']\n",
      "2025-04-02 15:33:46,067 - utils.utils - INFO - file processing - <_io.TextIOWrapper name='ner_config/embedding.yaml' mode='r' encoding='utf-8'>, type: embedder\n",
      "2025-04-02 15:33:46,068 - utils.utils - INFO - Trying config paths: ['ner_config/flow_ner.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/flow_ner.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/flow_ner.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/flow_ner.yaml']\n",
      "2025-04-02 15:33:46,071 - utils.utils - INFO - file processing - <_io.TextIOWrapper name='ner_config/flow_ner.yaml' mode='r' encoding='utf-8'>, type: flow\n",
      "2025-04-02 15:33:46,072 - utils.utils - INFO - Trying config paths: ['ner_config/search_ontology_knowledge.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/search_ontology_knowledge.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/search_ontology_knowledge.yaml', '/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/ner_example_ollama/ner_config/search_ontology_knowledge.yaml']\n",
      "2025-04-02 15:33:46,072 - utils.utils - INFO - file processing - <_io.TextIOWrapper name='ner_config/search_ontology_knowledge.yaml' mode='r' encoding='utf-8'>, type: knowledge\n",
      "2025-04-02 15:33:46,102 - chromadb.telemetry.product.posthog - INFO - Anonymized telemetry enabled. See                     https://docs.trychroma.com/telemetry for more information.\n",
      "\u001b[1;34mðŸŒŠ Flow: \u001b[0m\u001b[34mStructSenseFlow\u001b[0m\n",
      "\u001b[37m    ID: \u001b[0m\u001b[34m5931eff8-4423-41a8-a6e8-05e70ed871ee\u001b[0m\n",
      "â””â”€â”€ \u001b[33mðŸ§  Starting Flow...\u001b[0m\n",
      "\n",
      "\u001b[1m\u001b[35m Flow started with ID: 5931eff8-4423-41a8-a6e8-05e70ed871ee\u001b[00m\n",
      "2025-04-02 15:33:46,436 - crewai.flow.flow - INFO - Flow started with ID: 5931eff8-4423-41a8-a6e8-05e70ed871ee\n",
      "\u001b[1;34mðŸŒŠ Flow: \u001b[0m\u001b[34mStructSenseFlow\u001b[0m\n",
      "\u001b[37m    ID: \u001b[0m\u001b[34m5931eff8-4423-41a8-a6e8-05e70ed871ee\u001b[0m\n",
      "â”œâ”€â”€ \u001b[33mðŸ§  Starting Flow...\u001b[0m\n",
      "â””â”€â”€ \u001b[1;33mðŸ”„ Running:\u001b[0m\u001b[1;33m kickoff_flow\u001b[0m\n",
      "\n",
      "2025-04-02 15:33:46,438 - structsense.app - INFO - Running step: extracted_structured_information\n",
      "2025-04-02 15:33:47,202 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:33:47,960 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNeuroscience Named Entity Recognition (NER) Extractor Agent\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mFrom the given literature extract named entities from neuroscience statements.  A named entity is anything that can be referred to with a proper name.  Some common named entities in neuroscience articles are animal species (e.g., mouse, drosophila, zebrafish), anatomical regions (e.g., neocortex, mushroom body, cerebellum), experimental conditions (e.g., control, tetrodotoxin treatment, Scn1a knockout), and cell types (e.g., pyramidal neuron, direction-sensitive mechanoreceptor, oligodendrocyte)\n",
      "Literature: {'metadata': {'title': 'Independent Continuous Tracking of Multiple Agents in the Human Hippocampus', 'authors': ['Assia Chericoni', 'Justin M Fine', 'Ana G Chavez', 'Melissa C Franch', 'Elizabeth A Mickiewicz', 'Raissa K Mathura', 'Joshua Adkinson', 'Eleonora Bartoli', 'Joshua Jacobs', 'Nicole R Provenza', 'Andrew J Watrous', 'Seng Bum', 'Michael Yoo', 'Sameer A Sheth', 'Benjamin Y Hayden'], 'abstract': 'The pursuit of fleeing prey is a core element of many species behavioral repertoires. It poses the difficult problem of continuous tracking of multiple agents, including both self and others. To understand how this tracking is implemented neurally, we examined responses of hippocampal neurons while humans performed a joystick-controlled continuous prey-pursuit task involving two simultaneously fleeing prey (and, in some cases, a predator) in a virtual open field. We found neural maps encoding the positions of all the agents. All maps were multiplexed in single neurons and were disambiguated by the use of the population coding principle of semiorthogonal subspaces, which can facilitate cross-agent generalization. Some neurons, more common in the posterior hippocampus, had narrow tuning functions reminiscent of place cells, lower firing rates, and high information per spike; others, which were found in both anterior and posterior hippocampus, had broad tuning functions, higher firing rates, and less information per spike. Semi-orthogonalization was selectively associated with the broadly tuned neurons. These results suggest an answer to the problem of navigational individuation, that is, how mapping codes can distinguish different agents, and establish the neuronavigational foundations of pursuit.', 'publication_date': ''}, 'sections': [{'heading': 'INTRODUCTION', 'content': \"['Pursuit is a foraging behavior involving continuous and interactive navigation with the goal of catching a fleeing prey while avoiding predators (Fabian et al., 2018;Olberg et al., 2000;Stephens  Krebs, 1986;Ydenberg  Dill, 1986). It is an archetypal example of continuous decision-making, in which choice and control are deployed simultaneously (Burge et al., 2025;Cisek and Kalaska, 2010;Gordon et al., 2021;Merel et al., 2015;Yoo et al., 2021A). Moment to moment choices during pursuit require tracking the locations of multiple agents at the same time, including the self, pursued prey, and unpursued prey, and navigating towards interception loci.', 'Among regions associated with navigation, the hippocampus is the most well studied (Chersi and Burgess, 2015;Ekstrom et al., 2018;Epstein et al., 2017;Kunz et al., 2021;Maguire et al., 2006;Nyberg et al., 2022;OKeefe and Dostrovsky, 1971;Suthana et al., 2009). The hippocampus contains place cells that track the allocentric position of the self in physical space (Ekstrom et al., 2003;Jacobs et al., 2010;Miller et al., 2013;OKeefe et al., 1998;Wilson and McNaughton, 1993) and virtual space (Harvey et al., 2009;Mackay et al., 2024). The hippocampus also has a variety of other neurons relevant to pursuit (Behrens et al., 2018). These include neurons that are tuned to positions of external (physical) goals (Brown et al., 2016;Gauthier and Tank, 2018;Kunz et al., 2021;Poucet and Hok, 2017;Watrous et al., 2018), and social place cells that track the positions of other agents (Danjo et al., 2018;Forli and Yartsev, 2023;Omer et al., 2018;Rao et al., 2019;Stangl et al., 2021;Zang et al., 2024). The collective existence of these neurons indicates that the hippocampus contains the basic ingredients to track both self and prey in the course of pursuit. However, it does not give insight into how the brain solve the problem of disambiguating representations when faced with multiple distinct agents.', 'The need to distinguish self from others, and to distinguish multiple others, means that the brain must solve the individuation problem. In short, the brain needs a mechanism to know which agent a neurons responses refer to. One possibility would involve labelled line coding, in which separate sets of neurons track each single agent. However, such codes tend to be inflexible and have limited capacity for generalization (Barak et al., 2013;Fine et al., 2023;Fusi et al., 2016). In addition to these theoretical concerns, there are empirical ones: most brain areas contain mixed selective codes rather than labeled line codes (Ebitz and Hayden 2021;Fusi et al., 2016;Rigotti et al., 2013;Tye et al., 2024). One way for the brain to handle disambiguation of maps for distinct agents, despite the mixed selectivity, is to represent the maps in semi-orthogonal population subspaces (Elsayed et al., 2016;Johnston et al., 2024;Kaufman et al., 2022;Parthasarathy et al., 2017;Tang et al., 2020;Xie et al., 2022;Yoo and Hayden, 2020).', 'Here, we tested the hypothesis that the hippocampus uses subspace semi-orthogonalization to individuate maps related to distinct agents during continuous pursuit.', 'We recorded populations of neurons in the hippocampus in thirteen humans performing a virtual pursuit task with two prey and (in a subset of participants) a predator. We found neurons that encode position maps of the self, both prey, and the predator. Some neurons encoded the position of one agent, but the majority were mixed selective for multiple agents; in these mixed selective neurons, maps for different agents were largely unrelated. We found these hippocampal maps can be readily separated into two types: one with narrow spatial extent, reminiscent of place cells, and one with broader less-localized tuning. Narrowly tuned neurons were more prevalent in the posterior hippocampus and encoded position more efficiently, exhibiting greater spatial information per spike despite diminished firing rates, compared to broadly tuned neurons.', 'Notably, the population used subspace semi-orthogonalization for different agents; this coding principle was observed primarily in the broadly tuned neurons. Thus, rather than relying on labeled-line coding, the hippocampus appears to individuate multiple agents through populationlevel subspace organization, allowing for flexible encoding of distinct yet overlapping spatial representations.']\"}, {'heading': 'Prey pursuit in humans', 'content': \"['Human participants (n13) performed the prey-pursuit task (Figure 1A, Methods; Yoo et al., 2020). On each trial, the participant used a joystick to continuously move the position of an avatar (yellow circle) in a rectangular field displayed on a computer screen (Supplementary Video, Figure 1A andB). The participant had up to 20 seconds to capture fleeing prey (colored squares) to obtain points. Prey avoided the avatar with a deterministic strategy that combined repulsion from the avatars current position with repulsion from the walls of the field (Methods).', 'The prey items were drawn randomly on each trial from a set of three that differed in maximum velocity and reward size.', 'Each trial began with one or two prey appearing at one of the cardinal points (Figure 1A). In two-prey trials (81 of trials), the participant was free to decide which prey to pursue at any moment (Figure 1C,D). Participants successfully captured the prey in 73.72 of trials and, on successful trials, did so in an average of 7.51 seconds (variance: 1.67 seconds) with an average reaction time of 0.87 seconds (variance: 0.01 seconds), measured as the interval between the agents appearance on the screen and the participants first move. Participants performance did not depend significantly on prey type (trial length x reward level, p  0.16; reaction time x reward level, p  0.51). In three participants, we employed a variant of the task in which, in addition to the prey, there were also predators pursuing the participants avatar (Figure 1E). In this case, the predator used a simple distance-minimizing pursuit strategy. Our participants successfully evaded capture by the predator in 91.67 of trials.', 'We recorded responses of 390 neurons in the hippocampus while participants performed this task (average n30 neurons per participant). Of these neurons, 96 were also recorded in the variant of the task with a predator. Of all 390 neurons, roughly half (n199) were in anterior hippocampus and the remainder (n191) were in posterior hippocampus (Figure 1F). We defined the border between these regions as a coronal plane along the longitudinal hippocampal axis (y  -20, MNI). This border is largely consistent with that used in previous studies (e.g., Poppenk et al., 2013). Average firing rates aligned to trial stop and start show some intriguing patterns (Figure 1G). Overall, however, neurons had complex selectivities that were not readily explainable in terms of trial start and stop. We therefore examined responses as a function of location of each agent in the virtual space of the task. Participants use a joystick to control the position of an avatar (yellow circle) on a computer screen to capture prey (squares) and score points. For some participants, we also included predators (Methods). B, Example sessions from two participants; blue lines indicate participants trajectories on each trial overlaid; red dots indicate points of prey capture. C, D, E, Typical example trials from different participants. Trials are identified as participant number.trial number. Continuous gray lines indicate the chosen prey trajectories and dashed gray lines the unchosen prey trajectories. Predators trajectories are reported in magenta. D, Example trial in which the participant switched from pursuing one prey to another. E, Example of trial with predator -capture from the predator leads to point loss. F, Recording sites of hippocampal neurons from all 13 participants. Recording sites within the anterior hippocampus are reported in magenta while recording sites within the posterior hippocampus in purple. G, Peri-stimulus time histograms and raster plots demonstrating responses to chase start and chase end from two hippocampal neurons from participant 2 (left column) and participant 9 (right column). Dashed vertical lines represent the beginning and the end of the chase. Plots display mean firing rates.']\"}, {'heading': 'Hippocampal maps for positions of self, prey, and predators', 'content': \"['To estimate mapping functions in these neurons, we used the Poisson generalized linear model procedure developed by Hardcastle et al. (2017). This approach fits tuning models to neuronal responses without any a priori assumptions about the shape of the tuning surface. For this analysis, and all the subsequent ones, we concatenated all the successful trials involving two prey. (Single prey trials showed similar results and are not described here). To describe our results here, we use the term chosen prey for the prey that was ultimately captured, while the other was the unchosen prey (Figure 1C,D).', 'We found 37.7 (n147390) of neurons map the position of the self, while 33.9 (n132390) map the position of the chosen prey, and 24.4 (n95390) map the position of the unchosen prey (Figure 2A,B,D,F,G). In the neurons that we recorded during predator trials, 27.08 (n2696) map the position of the predator (Figure 2E). We found that 15.1 (n59390) of the neurons encode the position of any two agents, and 14.4 (n56390) map all three agents (Figure 2B). Similarly, during predator trials, 15.6 of the neurons (n1596) map the position of two or three agents, with 8.3 (896) mapping the position of all four agents. These proportions are all higher than would be expected by chance (p0.001, binomial test).', 'Next, we asked whether hippocampal neurons have different maps for the different agents. To quantify the relationship between maps, we used a spatial similarity index (SPAEF, Koch et al., 2018, Methods), which measures the correlation between spatial representations. A value of zero indicates full orthogonality between maps, while values closer to 1 or -1, indicate correlation or anti-correlation between maps, respectively.', 'The mean spatial similarity between self and chosen prey maps is 0.14. This value is very low, but is nonetheless greater than zero (p  0.002, Figure 2H, Methods) and below noise ceiling (p  0.001). Thus, these maps appear to be largely, but not entirely distinct. The mean spatial similarity between self and unchosen prey maps is even lower, but is still different from zero (SPAEF  0.002, p  0.001, Figure 2H). The chosen and unchosen prey maps are weakly, but significantly anti-correlated (SPAEF  -0.02, p  0.001, Figure 2H). Also in this case, both values were below noise ceiling (p  0.001). Finally, in the subset of neurons in which we had predator data, we found that the spatial similarity between self and predator was again slightly positive but not different from zero (SPAEF  0.006, p  0.79, Figure 2I). These single-neuron SPAEF results are reminiscent of the idea of semi-collinearity, in which coding at the population level shows a mixture of orthogonality and collinearity (Johnston et al., 2024). We therefore next tested this population level idea directly.', 'Figure 2. Hippocampal mapping functions for self and prey. A, Bar plot showing the percentage of neurons tuned to agent positions according to the LN-GLM approach. In total 51.79 (202390) of the neurons were selective for any of the agents positions. Proportions do not sum up to hundred as one neuron may be tuned to more than one agents position. Dashed horizontal grey line represents chance level, which is 5. Error bars represent standard errors. Significance was determined using a chi-square test for proportions, with a significance threshold of 0.05. B, Barplot showing the proportion of neurons tuned to more than one agents position. For instance, the first bar from the left shows that 14.36 of the neurons were selective for the position of all three agents. C, Quantification of model performance using loglikelihood (LLH) increase. The histogram shows LLH increase values for neurons significantly tuned to at least one agents position. LLH increase was computed as the improvement over a null model assuming a constant firing rate, using 10-fold cross-validation. Only neurons where the best-fit model significantly outperformed the null model (Wilcoxon signed-rank test, p  0.05) are shown. LLH increase values were normalized by spike count and converted to bits per spike using log base 2 scaling. D, E, F, G, Three-dimensional (top row) and two-dimensional (bottom rows) representations of spatial firing rate maps for neurons significantly tuned to the position of all the agents. Each row corresponds to a single neuron, while each column represents the firing activity relative to different agents (self, chosen prey, unchosen prey, and predator when present). Yellower regions indicate locations where the neuron exhibited higher firing rates. Data points were interpolated for visualization purposes only. H, Boxplots representing the median spatial similarity between maps across different agent representations, errorbars represent the standard error. Spatial similarity values near zero indicate orthogonality (low similarity) between maps. All the neurons are shown. J, Example maps from neurons that were significantly tuned to self position.']\"}, {'heading': 'Narrow and broad tuning curves for spatial position', 'content': \"['We next surveyed the distribution of tuning curve shapes. Some hippocampal neurons have selectivity for positions that are localized to a specific location, akin to the narrow maps of place cells (Figure 2D,E,G). However, others have broader response functions that are not as narrowly localized, but that nonetheless carry strong spatial information (Figure 2F,J).', 'We performed a k-means clustering of self-position tuning functions (Methods). We validated the clustering output using silhouette scores, which can provide an estimate of the most likely number of true clusters (Rousseeuw, 1987). This analysis shows that only two clusters are needed (mean silhouette value  0.6, Figure 3A,B). In our population of neurons, 70.5 (n275390) belong to cluster 1 (C1) which has neurons with broad spatial tuning; the remaining 29.5 (n115390) are in cluster 2 (C2), which has neurons with narrow place cell-like tuning (Figure 3A,B).', 'We applied principal component analysis (PCA) on the self-position tuning functions and visualized the population distribution in a lower dimensional space (Methods). Neurons belonging to cluster 1 and cluster 2 are clearly separable along the first principal component (PC1), which represents the overall spatial extent of the tuning, while the second principal component (PC2) may capture differences in neurons preferred firing locations, signifying spatial preference (Figure 3C). Indeed, neurons with broad spatial representations are distributed on the right side of PC1, with little variability along both PC1 and PC2, which may indicate greater spatial extent and lower spatial preference. That is, firing patterns are more widespread and heterogeneous across multiple spatial locations (Figure 3C). In contrast, neurons with narrow spatial representations were more dispersed along both PC1 and PC2, reflecting greater variability in their tuning peaks -spatial extent -and spatial preference (Figure 3C).', 'We performed the same clustering procedure, but this time using chosen preys position tuning functions (Methods). As with self-position, the silhouette test showed optimal results with k  2 (mean silhouette value  0.5, Figure 3B), confirming that population response relies on broad and narrow maps to represent position (Figure 3D). For these chosen prey maps, we found that 57.44 (n224390) belong to cluster 1, while the remaining 42.56 (n166390) belong to cluster 2. The overwhelming majority of neurons (79.80) consistently fall into the same cluster for self-and prey-position (Figure 3E,F,G,I), whereas the remaining 20.20 of neurons switch between broad and narrow spatial maps (Figure 3E,F,H,J). Repeating the same analysis on both self and chosen prey maps, first applying PCA followed by k-means clustering, yielded similar results.', 'We found that broad and narrow neuron maps are differentially distributed along the longitudinal axis of the hippocampus. Specifically, we observed a greater preponderance of broadly tuned neurons in the anterior hippocampus and a greater preponderance of neurons with narrow spatial tuning in the posterior hippocampus (Figure 3K). Among cluster 1 neurons, 53.80 localized rostrally with the remaining 46.20 caudally, with no significance difference between them (one sample z-test for proportions: p  0.99). Among cluster 2 neurons, 44.30 are found rostrally and 55.70 caudally. While this difference is modest, it is statistically significant (one sample z-test for proportions: p  0.007; Figure 3K). This result remained unchanged even when considering only the significantly tuned neurons. Out of the cluster 2 tuned neurons, 42.10 localize in the anterior hippocampus and 57.90 in the posterior (one sample z-test for proportions: p  0.003; data not shown). Moreover, within the posterior hippocampus, we found a higher proportion of cluster 2 neurons compared to cluster 1 (one sample z-test for proportions: p  0.021; Figure 3K). Within the anterior hippocampus, instead, there was no significant difference between the proportion of cluster 1 and cluster 2 neurons (one sample z-test for proportions: p  0.98, Figure 3K). Together, these results confirm theories that anterior and posterior hippocampus have measurable functional differences.', 'Last, we examined how these two distinct categories of neurons fire across space along with the associated signal-to-noise ratio (SNR), defined as the tuning curve variance normalized by firing rate (Methods). We found that neurons carrying broad spatial representation exhibit higher firing rates but carry less spatial information compared to narrowly tuned neurons (ranksum, p  0.001; Figure 3L,M). Nonetheless, the SNR remained comparable across clusters (Figure 3N), indicating that the firing rate and information per spike roughly cancelled each other out. However, we did find that neurons in the posterior hippocampus had lower firing rates but higher SNR compared to anterior hippocampal neurons (Figure 3L,N). These findings highlight distinct encoding strategies along the longitudinal hippocampal axis.', 'within neurons between 0 and 1. B, Silhouette scores used to assess clustering quality and determine the optimal number of clusters for both self and chosen prey maps. Stars indicate the selected number of clusters. C, Neurons projected onto the PC space derived from self-position maps, colored according to k-means clustering performed on self maps. D, Same as in C, but PCA and k-means were performed on chosen prey maps. E, Neurons projected onto the chosen prey PCs space, now colored by cross-referencing k-means clustering results from self and chosen prey maps. F, Proportion of neurons assigned to each cluster based on their self and chosen prey representation. The table highlights the percentage of neurons that remained in the same cluster across self and chosen prey representations, as well as those that switched clusters between conditions. G, H, I, J, Example neurons demonstrating shared (G, I) or distinct (H, J) spatial representations for self and chosen prey. Each column corresponds to a single neuron, with the top rows showing spatial tuning to self and the bottom row tuning to chosen prey. K, Comparison of neuron distribution between anterior and posterior hippocampus across clusters. Significance is based on one sample z-test for proportions. L, Plot representing the mean firing rate across a 6x6 grid (36 bins -x axis). All p  0.001, Wilcoxon rank sum test. M, Boxplot representing the spatial information per spike computed as the mutual information between the spatial location and the spike train, p  0.001, Wilcoxon rank sum test. N, Boxplot representing the SNR computed as the ratio between the variance of the tuning and the average firing rate, where a higher SNR means a neuron fires more reliably in specific locations; p  0.02, Wilcoxon rank sum test.']\"}, {'heading': 'Hippocampal neurons use semi-orthogonal maps to disambiguate agents', 'content': \"['Theoretical and empirical data support the idea that neural populations can use orthogonal neural subspaces to partition information encoded in the same neurons (Ebitz and Hayden, 2021;Elsayed et al., 2016;Kaufman et al., 2022;Panichello and Buschman, 2021;Tang et al., 2020;Xie et al., 2022;Yoo and Hayden, 2020). Moreover, blending orthogonal and collinear subspaces, producing semi-orthogonal subspaces, allows for discrimination while enabling cross-categorical generalization (Figure 4A-D; Barak et al., 2013;Bernardi et al., 2020;Johnston et al., 2024).', 'To test the hypothesis that hippocampal neural populations use semi-orthogonal subspaces to disentangle agents maps, we adapted a method previously developed to study motor regions (Elsayed et al., 2016). This analysis quantifies the degree of overlap between spatial representations for self and prey, by projecting the population activity onto lowdimensional subspaces that capture most of the variance between agents representations (Methods). We applied this analysis separately to both clusters. We found that neurons with broad spatial maps (cluster 1) use distinct, semi-orthogonal subspaces to individuate agents locations. We also found that neurons with narrow spatial maps (cluster 2) represent self and prey on nearly collinear subspaces.', 'In more detail, we started by analyzing the correlation structure for self and prey representations. For each agent, we computed the spatial similarity between the tuning surfaces for each pair of neurons (Figure 4E,F). Each entry in the resulting spatial similarity matrix represents the degree of spatial similarity between the response patterns of two neurons. For C1 neurons, the matrix structure changes markedly across agents (Figure 4E), whereas for C2 neurons, it remains largely consistent (Figure 4F). Indeed, when we compared the entries of the self matrix against the entries of the prey matrix, we found little or no correlation (C1: not different from chance, determined by pseudocorrelation, self vs chosen prey: R2  0.19  0.040, p  0.46; self vs unchosen prey: R2  0.06  0.039, p  0.99; two-sided permutation test, Î±  0.05; Figure 4G; C2: self vs chosen prey: RÂ²  0.19  0.037, p  0.001; self vs unchosen prey: RÂ²  0.14  0.038, p  0.006, Figure 4H). Thus, for a pair of neurons with broad spatial maps (cluster 1), the similarity of their responses to self-locations does not inform the similarity of their responses to prey locations. A possibility is that these neurons individuate agents in space with some dimensions dedicated to self-only, some to prey-only, and some to both -aligning with the hypothesis of semi-orthogonality between neural subspaces.', 'To understand if these subspaces were semi-orthogonal, we applied eigendecomposition to the self spatial similarity matrix and individuated the top self-principal components (PCs). If self and prey maps are semi-orthogonal, the self-PCs should capture very little prey variance (Methods). To assess if self-PCs can significantly capture the prey maps variance, we computed variance explain noise ceiling and confidence intervals through half-split cross-validation (CV) (Methods). For each split, we used half of the self spatial maps (train half) to compute the self-PCs. We then use these self-PCs to project the test halfs spatial maps for both self (within-self variance) and prey (cross-agents variance). The distribution of within-self variances gives an estimate of noise ceiling, while the cross-agents variance explained quantifies how well self-PCs capture prey maps variance. The variability of both within-self and across-agents variance provides confidence intervals for each agents projections.', 'For cluster 1 neurons, we found that the prey variance explained was significantly lower than noise ceiling (p  0.001, for PC1 to PC10), though confidence intervals did not overlap between self and prey projections for PC1 only. That is, self-PC1 carries information unique to the self maps. Overlapping confidence intervals for the remaining components implied that self and prey representations share non-zero variance (Figure 4I).', 'Lastly, to estimate the degree of orthogonality between agents neural subspaces, we identified the top prey-PCs, and estimated the cosine similarity between self and prey -PCs. A value of cosine similarity close to zero means orthogonality, while values closer to 1 or -1 collinearity. The cosine similarity values between the self-PCs and prey-PCs, fluctuate around zero, confirming weak alignment between PCs, i.e., semi-orthogonality between subspaces (Figure 4K). In summary, these results tell us that neurons with broad spatial tuning individuate prey and self using semi-orthogonal neural subspaces.', 'The same decomposition analysis performed on neurons with sparse spatial tuning (cluster 2), yielded slightly different results. Given the significant correlations between self and prey representations, a potential explanation is that agents lie on collinear subspaces. We found that also in this case, the prey variance explained is significantly lower than noise ceiling (p  0.001, for PC1 to PC10), but confidence intervals overlap across all the PCs, indicating shared variance between self and prey maps (Figure 4J). The cosine similarity between the top two self-PCs and chosen prey-PCs significantly differs from zero (PCs1: 0.78, p  0.02; PCs2: -0.36, p  0.04, Figure 4L). Our results indicate that neurons with narrow spatial maps represent self and prey in nearly collinear subspaces, and that subspace semi-orthogonalization in hippocampal neurons is primarily driven by neural populations with broad spatial tuning. uncorrelated activity results in subspace orthogonalization (bottom row) and vice versa (upper row). D, Blue represents the self subspace, while green represents the prey subspace. Correlated neural activity results in collinear subspaces (top row) and viceversa (bottom row). E, Spatial similarity matrices for all the possible neuron pairs in cluster 1 for self and prey. Neurons of each matrix are sorted using a hierarchical clustering algorithm applied on the self matrix only. Z axis indicates the strength of correlation between the overall average activities of neurons pairs. Lack of obvious structure in the self matrix reflects strong functional reorganization between self and prey. F, Same as in E, but for cluster 2 neurons. G, Pairwise correlation for each pair of neurons in cluster 1 for self and prey. The lack of correlation indicates that the relationship between neurons changes across agents. Reported values indicate Pearsons correlation coefficient with 95 confidence intervals, and permutation-based p-values were computed using a two-tailed test, with significance considered at p0.05. H, Same as in E, but significant correlations indicate that the relationship between neuron pairs in cluster 2 is preserved across agents, suggesting a shared representational structure. I, Percent variance explained by each of self-PCs for prey maps projected onto the self. The low height of the green and grey bars relative to the blue ones illustrates the poor match between self and prey subspaces. J, As in I, but for cluster 2 neurons. K, Cosine similarity between self and prey PCs, values closer to zero indicate orthogonality between PCs. Filled dots represent nonzero values (two-tailed test, with significance considered at p0.05). L, as in K, but for cluster 2 neurons.']\"}, {'heading': 'DISCUSSION', 'content': \"['Here we find, in a virtual prey pursuit task, that neurons in the human hippocampus can track positions of multiple agents (self, two prey, and a predator) simultaneously using distinct maps. Rather than use a labelled line code (in which a given neurons responses correspond to the position of a single agent), maps are multiplexed. This multiplexing of information raises an individuation problem: if firing rate of a neuron changes, which of the tracked agents does that change refer to? Our data suggest that the hippocampus solves this problem by recourse to semiorthogonal subspaces, which blend orthogonal and collinear subspaces. Subspace partition allows for simultaneous separation of information (thus solving the individuation problem) while also allowing for cross-agent generalization, greatly facilitating flexible learning (Barak et al., 2013;Fusi et al., 2016;Johnston et al., 2024). These results, then, build on recent findings showing that rat and bat hippocampus contains cells whose firing rates encode the positions of conspecifics in the environment. Specifically, we find that cells with similar functional properties can be found in the human hippocampus, and moreover, that they can flexibly partition information to achieve individuation.', 'We find evidence for two categories of cells, based on the shape of their tuning surface for location (broad and narrow). The cells in the smaller cluster 2 have narrow highly localized regions of strong activity, which give them a clear featural resemblance to place cells. However, the cells in the larger cluster 1 have a distinct lack of localized tuning, but nonetheless still carry information about spatial position. These cells have some resemblance to non-grid cells in the rat entorhinal cortex, which were discovered using the same mapping algorithm we employ here (Hardcastle et al., 2017 see also Diehl et al., 2017). Moreover, the category a cell falls in for selfmaps is generally the same category as it falls in for other maps, suggesting map shape may be an intrinsic property of neurons, rather than an arbitrary feature. Surprisingly, the subspace semiorthogonalization appears mostly to be a property of the cluster 1 neurons, suggesting that an overly strict focus on place-cells with focal tuning may result in ignoring important functional features of hippocampal maps.', 'Our results show a modest but highly significant difference in the distribution of functional cell types in the anterior (fewer cluster 2) and posterior (more cluster 2) hippocampus.', 'These results relate to ongoing discussions about the potential functional differences between these regions (which are homologous to the rodent ventral and dorsal hippocampi, respectively, Strange et al., 2014;To et al., 2024). The increased number of place-cell like coding in posterior hippocampus, then, recapitulates known findings from rats (e.g., Jung et al., 1994;Kjelstrup et al., 2018). Finally, the localization of partitioning to the more anterior cluster 1 neurons aligns this part of the hippocampus with the kind of flexible adaptive cognition associated with prefrontal cortex, with which anterior hippocampus has strong connections (Dalton et al., 2022).', 'Nonetheless, the differences we observe are modest, and, overall, our results endorse functional continuity of the human hippocampus.', 'Our results relate to a previous series of three papers from our laboratory, in which macaques performed a similar task (Yoo et al., 2020 and2021B;Fine et al., 2024). In those papers, we found that dorsal anterior cingulate cortex (dACC) actively predicts the future position of pursued prey about 700 ms into the future (Yoo et al., 2020), also tracks a suite of allocentric and egocentric variables related to the task (Yoo et al., 2021B), and corresponds with inferred decision variables in the task (Fine et al., 2024). Relative to macaques, human behavior is quite different -most critically, our macaque subjects needed at least three months to learn the task, whereas humans need no training. Moreover, humans show evidence of highly efficient behavioral strategies, such as cornering the prey, that macaques do not seem to learn, even with months of training. In any case, our results here, focusing on the individuation problem, which was not addressed in our earlier work, complement and extend our earlier findings.', 'We find it interesting that hippocampus actively tracks the position of the unpursued prey. In our task, subjects typically select a single prey and pursue it for several seconds until capture, leaving the other, unchosen one, ostensibly irrelevant. Nonetheless, from the perspective of the hippocampus, this unselected prey shows only modest evidence of deprioritization.', '(Namely, hippocampus still tracks the unpursued prey, although less robustly as it tracks the pursued prey). The goal of the task is not, however, simply to track the preferred prey; instead the player is allowed to switch to following the alternative (and sometimes does so, when, for example, the unpursued prey comes nearby). Thus, from the perspective of choice, although not behavior, monitoring the position of the unpursued prey is crucial: because switching is always a possibility, continuing to focus on the pursued prey represents a decision to not switch. Our results are reminiscent of theories in which the hippocampus maps potential alternatives to current goals, rather than being limited to representing only the goals themselves.']\"}, {'heading': 'Human intracranial neurophysiology', 'content': \"['Experimental data were recorded from 13 adult patients (6 males and 7 females) undergoing intracranial monitoring for epilepsy. The hippocampus was not a seizure focus area of any patients included in the study. Single neuron data were recorded from stereotactic (sEEG) probes, specifically AdTech Medical probes in a Behnke-Fried configuration. Each patient had an average of 3 probes terminating in left and right hippocampus. Electrode locations are verified by co-registered pre-operative MRI and post-operative CT scans. Each probe includes 8 microwires, each with 8 contacts, specifically designed for recording single-neuron activity. Single neuron data were recorded using a 512-channel Blackrock Microsystems Neuroport system sampled at 30 kHz. To identify single neuron action potentials, the raw traces were spike-sorted using the WaveClus sorting algorithm (Chaure et al., 2018) and then manually evaluated. Noise was removed and each signal was classified as multi or single unit using several criteria: consistent spike waveforms, waveform shape (slope, amplitude, trough-topeak), and exponentially decaying ISI histogram with no ISI shorter than the refractory period (1 ms). The analyses here used only single unit activity.']\"}, {'heading': 'Electrode visualization', 'content': \"['Electrodes were localized using the software pipeline intracranial Electrode Visualization (iELVis, Groppe et al., 2017) and plotted across patients on an average brain using Reproducible Analysis  Visualization of iEEG (RAVE, Magnotti et al., 2020). For each patient, DICOM images of the preoperative T1 anatomical MRI and the postoperative Stealth CT scans were acquired and converted to NIfTI format (Li et al., 2016). The CT was aligned to MRI space using FSL (Jenkinson and Smith, 2001;Jenkinson et al, 2002). The resulting coregistered CT was loaded into BioImage Suite (version 3.5Î²1) (Joshi et al., 2001) and the electrode contacts were manually localized.', 'Electrodes coordinates were converted to patient native space using iELVis MATLAB functions (Joshi et al., 2011;Yang et al., 2012) and plotted on the Freesurfer (version 7.4.1, Dale et al., 1999) reconstructed brain surface. Microelectrode coordinates are taken from the first (deepest) macro contact on the Ad-Tech Behnke Fried depth electrodes. RAVE was used to transform each patients brain and electrode coordinates into MNI152 average space (Magnotti et al., 2020). The average coordinates were plotted together on a glass brain with the hippocampus segmentation and colored by location within the hippocampus.']\"}, {'heading': 'The prey-pursuit task', 'content': \"['The task used here is similar but not identical to one we have previously used in macaques (Yoo et al., 2020A and2021;Fine et al., 2024; Supplementary Video). At the beginning of each trial, two shapes appeared on a gray background (RGB: 128128128) on a standard computer monitor placed in front of the subject. A yellow circle (15-pixels in diameter) was an avatar for the research participant. Its position was determined by the joystick and was limited by the screen boundaries. A square shape (30 pixels in length) represented the prey. The movement of the prey was determined by a simple artificial intelligence algorithm (see below). Each trial ended with either the successful capture of the prey or after 20 sec, whichever came first. Successful capture was defined as any spatial overlap between the avatar circle and the prey square.', 'Capture resulted in scored points; the number of points corresponded to prey color as follows: 1 point for orange; 3 points for green; 5 points for cyan.', 'The path of the prey was generated interactively using A-star pathfinding methods, which are commonly used in video gaming (Hart  Nils, 1968). For every frame (16.67 ms), we computed the cost of 15 possible future positions the prey could move to in the next time-step. These 15 positions were equally spaced on the circumference of a circle centered on the current position of the prey, with a radius equal to the maximum distance the prey could travel within one time-step. The cost in turn was based on the following two factors: the position in the field and the position of the avatar of the subject. The field that the prey moved in had a built-in bias for cost, which made the prey more likely to move toward the center. The cost due to distance from the avatar of the subject was transformed using a sigmoidal function: the cost became zero beyond a certain distance so that the prey did not move, and it became greater as distance from the avatar of the subject decreased. Eventually, the costs from these 15 positions were calculated and the position with the lowest cost was selected for the next movement. If the next movement was beyond the screen range (1,800  1,000 resolution), then the position with the second lowest cost was selected, and so on.', 'The maximum speed of the participant was 23 pixels per frame (and each frame was 16.67 ms). The maximum and minimum speeds of the prey remained the same across subjects. In trials with predators, a predator (triangle shape) appeared on 50 of trials. Capture by the predator led to points loss. Predators came in five different types (indicated by color) indicating different levels of points loss, ranging from 1 to 5 points.', 'The algorithm of the predator is to minimize the distance between itself and player.', 'Unlike the prey, the predator algorithm is governed by this single rule. The design of the task reflects primarily the desire to have a rich and variegated virtual world with opportunities for choices at multiple levels that is neither trivially simple nor overly complex.']\"}, {'heading': 'Task presentation', 'content': \"['Patients played at least 100 trials (average 119 trials) of the prey-pursuit task using a joystick (Figure 1A,B). The joystick was a commercially available joystick with a built-in potentiometer (Logitech Extreme Pro 3D). The joystick position was read out by a custom-coded program in Matlab running on the stimulus-control computer. The joystick was controlled by an algorithm that detected the positional change of the joystick and limited the maximum pixel movement to within 23 pixels in 16.67 ms. Task events were synchronized to the neural recording system via comments, sent through analog port, from the computer playing the task to the Neural Signal Processor (NSP) at 30 kHz.']\"}, {'heading': 'Linear-nonlinear model (LN-GLM).', 'content': \"['To test the selectivity of neurons for various experimental variables, we constructed GLMs with navigational variables (Hardcastle et al., 2017;Pillow et al., 2008). The GLMs estimated the spike rate of one neuron during time bin t as an exponential function of the weighted sum of the relevant value of each variable at time t, for which the weights are determined by a set of coefficients (wi). The estimated firing rates from the GLMs can be expressed as follows:', 'Where r denotes a vector of firing rates for one neuron over T time points across the session, and i indexes the variables of interest, for example, the position of the avatar on the screen. The vector of firing rates over T time points provide the benefit for modeling the neural activity without the need of specifically timelocking to a behavioral event. Xi is a matrix in which each column represents a set of state variables of the subject obtained from binning the continuous variable so that all the columns for a particular row are 0, except for one column. In this case, state variables were obtained binning the position of self and prey over a 6x6 grid (36 bins). As a result, each neuron was assigned a design matrix with rows corresponding to the number of samples and columns corresponding to the number of bins.', 'Unlike conventional tuning curve analysis, GLM analysis does not assume the parametric shape of the tuning curve a priori. Instead, the weights, which define the shape of tuning for each neuron, were optimized by maximizing the Poisson loglikelihood of the observed spike train given the model-expected spike number, with additional regularization for the smoothness of parameters in a continuous variable, and a lasso regularization for parameters in a discrete variable. Position parameters were separately smoothed across rows and columns. The regularization hyperparameter was chosen by maximizing the cross-validation log-likelihood based on several randomly selected neurons. The unconstrained optimization with gradient and Hessian was performed (Matlab fminunc function). The model performance of each neuron was quantified by the log-likelihood of held out data under the model (Figure 2C). This cross-validation procedure was repeated ten times (tenfold cross-validation), and overfitting was penalized. Through multiple levels of penalties, we compared the performance of models with varying complexity.']\"}, {'heading': 'Forward model selection', 'content': \"['Model selection was based on the cross-validated log-likelihood value for each model. We first fit n models with a single variable, where n is the total number of variables. The best single model was determined by the largest increase in spikenormalized log-likelihood from the null model (that is, the model with a single parameter representing the mean firing rate). Then, additional variables (n -1 in total) were added to the best single variable model. The best two variable model was preferred over the single variable model only if it significantly improved the cross-validation log-likelihood value (Wilcoxon signed-rank test, Î±  0.05). Likewise, the procedure was continued for the three-variable model and beyond if adding more variables significantly improved the model performance, and the best, simplest model was selected. The cell was categorized as not tuned to any of the variables considered if the log-likelihood increase was not significantly higher than baseline, which was the mean firing rate of fitted neurons across the session.']\"}, {'heading': 'Spatial similarity index (SPAEF)', 'content': \"['To compare the similarity between two spatial representations, we used the spatial efficiency measure (SPAEF) that prior literature suggests to be more robust than the 2D spatial correlation (Koch et al, 2018). It quantifies the similarity between two maps as follows:', 'where A is the Pearson correlation between two maps, B is the ratio between the coefficients of variation for each map and C is the activity similarity measured by histogram profiles. Values near -1 indicate anticorrelated maps (one tends to be high when the other is low), 0 indicates uncorrelated maps and 1 indicates perfect matching between the two. By definition, SPAEF is not strictly constrained between -1 and 1.', 'However, values outside this range tend to be rare, and in any case, in our data, values outside this interval never occurred.', 'To determine if the SPAEF values were significantly different from zero, we built a shuffled distribution over 1000 permutations. We then used a two-sided test to extract p-values, considering significance at p0.05. To determine the noise ceiling, we performed 1000 within-agent half-splits. That is, for each iteration and each neuron, we randomly split the trials for each agent into two independent halves, computed the spatial tuning maps separately for each half, and then measured the spatial similarity between the two resulting maps. This process estimates the upper bound of explainable variance by accounting for the inherent noise in the data, providing a reference against which observed spatial similarities can be compared.']\"}, {'heading': 'K-means clustering and Principal Component Analysis', 'content': \"['For this analysis, we included the entire neuronal population (390 neurons). We computed spatial tuning curves for each neuron based on the positions of the self and chosen prey. The task space (computer monitor) was divided into a 66 grid, resulting in a 36-bin tuning curve for each neuron per agent. Each tuning curve was then normalized between 0 and 1 on a per-neuron basis, by dividing by the maximum value of the respective neurons tuning curve. This ensured that differences in firing rates did not bias our subsequent analyses. We then applied the k-means clustering algorithm to the normalized tuning curves. Clustering was performed separately for self-position tuning curves and chosen prey-position tuning curves. To determine the optimal number of clusters we used silhouette values, which assess clustering quality by measuring how well each data point fits within its assigned cluster relative to other clusters. A silhouette score closer to -1 indicates potential misclassification, while a score closer to 1 suggests strong cluster cohesion and clear separation from other clusters (Rousseeuw, 1986). We selected the number of clusters that maximized the silhouette value, which turned out to be two for both self and chosen prey position tuning curves. Thus we could classify each neuron as belonging to cluster 1 or cluster 2.', 'We then performed principal component analysis (PCA) to visualize how the structure of the spatial tuning curves varies across neurons. That is, identifying the main patterns of variability in neurons responses to different positions. PCA was applied to the same normalized tuning curves (neurons  bins matrix) used for k-means clustering,', 'where each row represented a neuron and each column corresponded to a spatial bin in its tuning curve. The resulting principal components (PCs) captured different aspects of tuning variability across the population.', 'To visualize this structure, we plotted neurons in the PCA space using their PC1 and PC2 scores and colored them according to their k-means cluster assignment. This allowed us to assess whether neurons with similar tuning properties grouped together in the PCA-defined space.', 'Specifically, we applied PCA separately to the self-position tuning maps and the chosen prey tuning maps, coloring neurons according to their respective k-means clustering results (Figure 3C,D). Lastly, we re-applied PCA to the chosen prey tuning maps and colored neurons by cross-referencing their cluster identities from both self and chosen prey tuning maps (Figure 3E). This final analysis highlighted differences in cluster membership between self and chosen prey tuning representations.']\"}, {'heading': 'Information per spike and SNR', 'content': \"['This analysis was conducted on the raw (non-normalized) self-position spatial tuning curves (again 36 bins were used). Each neuron was assigned a cluster identity based on the results of k-means clustering applied to the self-position tuning curves. We calculated the spatial information per spike as the mutual information between the spatial location and the spike train, computed as:', 'Where ri is the firing rate of each spatial bin (36 bins total) for neuron n, r is the average firing rate across all bins, and pi is the probability of occupying each spatial bin (Skaggs et al., 1992). In our case, since participants were able to sample the whole task space (Figure 1B, C and Figure 3A), we assumed uniform occupancy.', 'To assess the stability of cells firing, we calculate the signal-to-noise ratio (SNR), defined as the variance in the mean firing rate across bins divided by the mean rate itself. This metric provides a measure of the consistency of spatial encoding (Fenton and Muller, 1998).']\"}, {'heading': 'ð‘†ð‘ð‘…  ð‘£ð‘Žð‘Ÿð‘–ð‘Žð‘›ð‘ð‘’ ð‘œð‘“ ð‘šð‘’ð‘Žð‘› ð‘¡ð‘¢ð‘›ð‘–ð‘›ð‘” ð‘ð‘¢ð‘Ÿð‘£ð‘’ ð‘šð‘’ð‘Žð‘› ð‘“ð‘–ð‘Ÿð‘–ð‘›ð‘” ð‘Ÿð‘Žð‘¡ð‘’', 'content': \"['Higher SNR signifies consistency in firing across spatial locations, while lower SNR indicates less spatially structured responses. Significance of differences between groups was assessed using a non-parametric Wilcoxon rank-sum test.']\"}, {'heading': 'Spatial similarity matrix', 'content': \"['To represent the correlation structure of self and prey representations, we constructed an NN matrix for each agent, with N number of neurons. Each matrix entry represents the spatial similarity between the spatial maps of neuron pairs. Spatial maps were normalized between 0 and 1. The spatial similarity matrix is symmetric and contains values between -1 and 1.']\"}]}\n",
      "\u001b[00m\n",
      "\u001b[92m15:33:48 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:33:48,013 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:35:20,451 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:35:20 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:35:20,565 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:35:20,615 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:35:20,616 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:35:20 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:35:20,633 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:35:20,682 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:35:20,704 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:36:46,441 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:36:46 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:36:46,534 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNeuroscience Named Entity Recognition (NER) Extractor Agent\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"extracted_terms\": {\n",
      "    \"1\": [\n",
      "      {\n",
      "        \"entity\": \"mouse\",\n",
      "        \"label\": \"ANIMAL_SPECIES\",\n",
      "        \"sentence\": \"These particles were visualized by fluorescent immunohistochemistry using mouse monoclonal anti-human myelin basic protein (MBPh) antibody (clone SMI-99).\",\n",
      "        \"start\": 79,\n",
      "        \"end\": 84,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"2\": [\n",
      "      {\n",
      "        \"entity\": \"oligodendrocyte\",\n",
      "        \"label\": \"CELL_TYPE\",\n",
      "        \"sentence\": \"Individual oligodendrocytes provide, on average, 16 near axons with single myelin segments about 200 Âµm in length (Butt and Ransom, 1993).\",\n",
      "        \"start\": 14,\n",
      "        \"end\": 29,\n",
      "        \"paper_location\": \"discussion\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"3\": [\n",
      "      {\n",
      "        \"entity\": \"cerebellum\",\n",
      "        \"label\": \"ANATOMICAL_REGION\",\n",
      "        \"sentence\": \"Myelin basic protein, human (MBPh) The mouse monoclonal antibody against the human myelin basic protein (clone SMI-99; Covance, Princeton, NJ) detects 4 bands between 14 and 21 kDa, corresponding to 4 myelin basic protein (MBP) isoforms on immunoblots of the mouse cerebellum (Dyer et al.).\",\n",
      "        \"start\": 293,\n",
      "        \"end\": 303,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "}\u001b[00m\n",
      "\n",
      "\n",
      "2025-04-02 15:36:46,557 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:36:46,589 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:36:46,691 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/.structsense_env/lib/python3.11/site-packages/chromadb/types.py:144: PydanticDeprecatedSince211: Accessing this attribute on the instance is deprecated, and will be removed in Pydantic V3. Instead, you should access this attribute from the model class. Deprecated in Pydantic V2.11 to be removed in V3.0.\n",
      "  return self.model_fields  # pydantic 2.x\n",
      "2025-04-02 15:36:46,738 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:36:46,751 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:36:46 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:36:46,756 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:38:21,387 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:38:21 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:38:21,434 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:38:21,473 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:38:21,473 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:38:21 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:38:21,482 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:38:21,507 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:39:55,360 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:39:55 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:39:55,409 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:39:55,446 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:39:55,447 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:39:55 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:39:55,454 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:39:55,485 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:41:28,950 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:41:28 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:41:28,998 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "Failed to add to long term memory: Failed to convert text into a Pydantic model due to error: 'NoneType' object has no attribute 'function_calling_llm'\n",
      "2025-04-02 15:41:29,028 - structsense.app - INFO - Running step: align_structured_information\n",
      "2025-04-02 15:41:29,060 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:41:29,095 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:41:29,276 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:41:29,347 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNeuroscience Named Entity Recognition (NER) Concept Alignment Agent\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mTake the output of extractor_agent {\n",
      "  \"extracted_terms\": {\n",
      "    \"1\": [\n",
      "      {\n",
      "        \"entity\": \"mouse\",\n",
      "        \"label\": \"ANIMAL_SPECIES\",\n",
      "        \"sentence\": \"These particles were visualized by fluorescent immunohistochemistry using mouse monoclonal anti-human myelin basic protein (MBPh) antibody (clone SMI-99).\",\n",
      "        \"start\": 79,\n",
      "        \"end\": 84,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"2\": [\n",
      "      {\n",
      "        \"entity\": \"oligodendrocyte\",\n",
      "        \"label\": \"CELL_TYPE\",\n",
      "        \"sentence\": \"Individual oligodendrocytes provide, on average, 16 near axons with single myelin segments about 200 Âµm in length (Butt and Ransom, 1993).\",\n",
      "        \"start\": 14,\n",
      "        \"end\": 29,\n",
      "        \"paper_location\": \"discussion\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"3\": [\n",
      "      {\n",
      "        \"entity\": \"cerebellum\",\n",
      "        \"label\": \"ANATOMICAL_REGION\",\n",
      "        \"sentence\": \"Myelin basic protein, human (MBPh) The mouse monoclonal antibody against the human myelin basic protein (clone SMI-99; Covance, Princeton, NJ) detects 4 bands between 14 and 21 kDa, corresponding to 4 myelin basic protein (MBP) isoforms on immunoblots of the mouse cerebellum (Dyer et al.).\",\n",
      "        \"start\": 293,\n",
      "        \"end\": 303,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "} as input and perform the concept alignment using the ontological concepts.  A concept alignment is anything where you align the given entity to the matching concept aka class from a ontology or a schema.\n",
      "\u001b[00m\n",
      "\u001b[92m15:41:29 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:41:29,365 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:43:02,934 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:43:02 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:43:02,976 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "\u001b[92m15:43:02 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:43:02,983 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:43:03,017 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:43:03,037 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:44:22,403 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:44:22 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:44:22,430 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNeuroscience Named Entity Recognition (NER) Concept Alignment Agent\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "{\n",
      "  \"extracted_terms\": {\n",
      "    \"1\": [\n",
      "      {\n",
      "        \"entity\": \"mouse\",\n",
      "        \"label\": \"ANIMAL_SPECIES\",\n",
      "        \"sentence\": \"These particles were visualized by fluorescent immunohistochemistry using mouse monoclonal anti-human myelin basic protein (MBPh) antibody (clone SMI-99).\",\n",
      "        \"start\": 79,\n",
      "        \"end\": 84,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"2\": [\n",
      "      {\n",
      "        \"entity\": \"oligodendrocyte\",\n",
      "        \"label\": \"CELL_TYPE\",\n",
      "        \"sentence\": \"Individual oligodendrocytes provide, on average, 16 nearby axons with single myelin segments about 200 Âµm in length (Butt and Ransom, 1993).\",\n",
      "        \"start\": 14,\n",
      "        \"end\": 29,\n",
      "        \"paper_location\": \"discussion\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"3\": [\n",
      "      {\n",
      "        \"entity\": \"cerebellum\",\n",
      "        \"label\": \"ANATOMICAL_REGION\",\n",
      "        \"sentence\": \"Myelin basic protein, human (MBPh) The mouse monoclonal antibody against the human myelin basic protein (clone SMI-99; Covance, Princeton, NJ) detects 4 bands between 14 and 21 kDa, corresponding to 4 myelin basic protein (MBP) isoforms on immunoblots of the mouse cerebellum (Dyer et al.).\",\n",
      "        \"start\": 293,\n",
      "        \"end\": 303,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "}\u001b[00m\n",
      "\n",
      "\n",
      "2025-04-02 15:44:22,454 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:44:22,477 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:44:22,544 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/.structsense_env/lib/python3.11/site-packages/chromadb/types.py:144: PydanticDeprecatedSince211: Accessing this attribute on the instance is deprecated, and will be removed in Pydantic V3. Instead, you should access this attribute from the model class. Deprecated in Pydantic V2.11 to be removed in V3.0.\n",
      "  return self.model_fields  # pydantic 2.x\n",
      "2025-04-02 15:44:22,586 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:44:22,603 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:44:22 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:44:22,608 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:45:17,364 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:45:17 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:45:17,385 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:45:17,415 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:45:17,416 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:45:17 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:45:17,428 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:45:17,479 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:46:11,289 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:46:11 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:46:11,306 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:46:11,326 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:46:11,327 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:46:11 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:46:11,337 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:46:11,359 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:47:05,086 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:47:05 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:47:05,104 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "Failed to add to long term memory: Failed to convert text into a Pydantic model due to error: 'NoneType' object has no attribute 'function_calling_llm'\n",
      "2025-04-02 15:47:05,115 - structsense.app - INFO - Running step: judge_alignment\n",
      "2025-04-02 15:47:05,125 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:47:05,143 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:47:05,317 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:47:05,494 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNeuroscience Named Entity Recognition (NER) Judge Agent\u001b[00m\n",
      "\u001b[95m## Task:\u001b[00m \u001b[92mTake the output of alignment agent {\n",
      "  \"extracted_terms\": {\n",
      "    \"1\": [\n",
      "      {\n",
      "        \"entity\": \"mouse\",\n",
      "        \"label\": \"ANIMAL_SPECIES\",\n",
      "        \"sentence\": \"These particles were visualized by fluorescent immunohistochemistry using mouse monoclonal anti-human myelin basic protein (MBPh) antibody (clone SMI-99).\",\n",
      "        \"start\": 79,\n",
      "        \"end\": 84,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"2\": [\n",
      "      {\n",
      "        \"entity\": \"oligodendrocyte\",\n",
      "        \"label\": \"CELL_TYPE\",\n",
      "        \"sentence\": \"Individual oligodendrocytes provide, on average, 16 nearby axons with single myelin segments about 200 Âµm in length (Butt and Ransom, 1993).\",\n",
      "        \"start\": 14,\n",
      "        \"end\": 29,\n",
      "        \"paper_location\": \"discussion\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"3\": [\n",
      "      {\n",
      "        \"entity\": \"cerebellum\",\n",
      "        \"label\": \"ANATOMICAL_REGION\",\n",
      "        \"sentence\": \"Myelin basic protein, human (MBPh) The mouse monoclonal antibody against the human myelin basic protein (clone SMI-99; Covance, Princeton, NJ) detects 4 bands between 14 and 21 kDa, corresponding to 4 myelin basic protein (MBP) isoforms on immunoblots of the mouse cerebellum (Dyer et al.).\",\n",
      "        \"start\": 293,\n",
      "        \"end\": 303,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "} as input and perform the following evaluation:  1. Assess the quality and accuracy of the alignment with the ontology or schema in {\n",
      "  \"extracted_terms\": {\n",
      "    \"1\": [\n",
      "      {\n",
      "        \"entity\": \"mouse\",\n",
      "        \"label\": \"ANIMAL_SPECIES\",\n",
      "        \"sentence\": \"These particles were visualized by fluorescent immunohistochemistry using mouse monoclonal anti-human myelin basic protein (MBPh) antibody (clone SMI-99).\",\n",
      "        \"start\": 79,\n",
      "        \"end\": 84,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"2\": [\n",
      "      {\n",
      "        \"entity\": \"oligodendrocyte\",\n",
      "        \"label\": \"CELL_TYPE\",\n",
      "        \"sentence\": \"Individual oligodendrocytes provide, on average, 16 nearby axons with single myelin segments about 200 Âµm in length (Butt and Ransom, 1993).\",\n",
      "        \"start\": 14,\n",
      "        \"end\": 29,\n",
      "        \"paper_location\": \"discussion\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"3\": [\n",
      "      {\n",
      "        \"entity\": \"cerebellum\",\n",
      "        \"label\": \"ANATOMICAL_REGION\",\n",
      "        \"sentence\": \"Myelin basic protein, human (MBPh) The mouse monoclonal antibody against the human myelin basic protein (clone SMI-99; Covance, Princeton, NJ) detects 4 bands between 14 and 21 kDa, corresponding to 4 myelin basic protein (MBP) isoforms on immunoblots of the mouse cerebellum (Dyer et al.).\",\n",
      "        \"start\": 293,\n",
      "        \"end\": 303,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "}. 2. Assign a score between 0 and 1 as a judge_score. 3. Update the {\n",
      "  \"extracted_terms\": {\n",
      "    \"1\": [\n",
      "      {\n",
      "        \"entity\": \"mouse\",\n",
      "        \"label\": \"ANIMAL_SPECIES\",\n",
      "        \"sentence\": \"These particles were visualized by fluorescent immunohistochemistry using mouse monoclonal anti-human myelin basic protein (MBPh) antibody (clone SMI-99).\",\n",
      "        \"start\": 79,\n",
      "        \"end\": 84,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"2\": [\n",
      "      {\n",
      "        \"entity\": \"oligodendrocyte\",\n",
      "        \"label\": \"CELL_TYPE\",\n",
      "        \"sentence\": \"Individual oligodendrocytes provide, on average, 16 nearby axons with single myelin segments about 200 Âµm in length (Butt and Ransom, 1993).\",\n",
      "        \"start\": 14,\n",
      "        \"end\": 29,\n",
      "        \"paper_location\": \"discussion\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"3\": [\n",
      "      {\n",
      "        \"entity\": \"cerebellum\",\n",
      "        \"label\": \"ANATOMICAL_REGION\",\n",
      "        \"sentence\": \"Myelin basic protein, human (MBPh) The mouse monoclonal antibody against the human myelin basic protein (clone SMI-99; Covance, Princeton, NJ) detects 4 bands between 14 and 21 kDa, corresponding to 4 myelin basic protein (MBP) isoforms on immunoblots of the mouse cerebellum (Dyer et al.).\",\n",
      "        \"start\": 293,\n",
      "        \"end\": 303,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "} adding the judge_score.\n",
      "\u001b[00m\n",
      "\u001b[92m15:47:05 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:47:05,521 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:48:56,040 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:48:56 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:48:56,074 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "\u001b[92m15:48:56 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:48:56,082 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:48:56,116 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:48:56,178 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:50:28,564 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/generate \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:50:28 - LiteLLM:INFO\u001b[0m: utils.py:1084 - Wrapper: Completed Call, calling success_handler\n",
      "2025-04-02 15:50:28,600 - LiteLLM - INFO - Wrapper: Completed Call, calling success_handler\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[95m# Agent:\u001b[00m \u001b[1m\u001b[92mNeuroscience Named Entity Recognition (NER) Judge Agent\u001b[00m\n",
      "\u001b[95m## Final Answer:\u001b[00m \u001b[92m\n",
      "```json\n",
      "{\n",
      "  \"extracted_terms\": {\n",
      "    \"1\": [\n",
      "      {\n",
      "        \"entity\": \"mouse\",\n",
      "        \"label\": \"ANIMAL_SPECIES\",\n",
      "        \"sentence\": \"These particles were visualized by fluorescent immunohistochemistry using mouse monoclonal anti-human myelin basic protein (MBPh) antibody (clone SMI-99).\",\n",
      "        \"start\": 79,\n",
      "        \"end\": 84,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"2\": [\n",
      "      {\n",
      "        \"entity\": \"oligodendrocyte\",\n",
      "        \"label\": \"CELL_TYPE\",\n",
      "        \"sentence\": \"Individual oligodendrocytes provide, on average, 16 near axons with single myelin segments about 200 Âµm in length (Butt and Ransom, 1993).\",\n",
      "        \"start\": 14,\n",
      "        \"end\": 29,\n",
      "        \"paper_location\": \"discussion\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ],\n",
      "    \"3\": [\n",
      "      {\n",
      "        \"entity\": \"cerebellum\",\n",
      "        \"label\": \"ANATOMICAL_REGION\",\n",
      "        \"sentence\": \"Myelin basic protein, human (MBPh) The mouse monoclonal antibody against the human myelin basic protein (clone SMI-99; Covance, Princeton, NJ) detects 4 bands between 14 and 21 kDa, corresponding to 4 myelin basic protein (MBP) isoforms on immunoblots of the mouse cerebellum (Dyer et al.).\",\n",
      "        \"start\": 293,\n",
      "        \"end\": 303,\n",
      "        \"paper_location\": \"methods\",\n",
      "        \"paper_title\": \"Concentration of myelin debris-like myelin basic protein-immunoreactive particles in the distal (anterior)-most part of the myelinated region in the normal rat optic nerve\",\n",
      "        \"doi\": \"10.1101/2025.03.19.643597\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "}\n",
      "```\u001b[00m\n",
      "\n",
      "\n",
      "2025-04-02 15:50:28,629 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:50:28,650 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:50:28,771 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/embeddings \"HTTP/1.1 200 OK\"\n",
      "/Users/tekrajchhetri/Documents/brainypedia_codes_design/crew_ner_framework/structsense/example/.structsense_env/lib/python3.11/site-packages/chromadb/types.py:144: PydanticDeprecatedSince211: Accessing this attribute on the instance is deprecated, and will be removed in Pydantic V3. Instead, you should access this attribute from the model class. Deprecated in Pydantic V2.11 to be removed in V3.0.\n",
      "  return self.model_fields  # pydantic 2.x\n",
      "2025-04-02 15:50:28,814 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "2025-04-02 15:50:28,833 - httpx - INFO - HTTP Request: POST http://localhost:11434/api/show \"HTTP/1.1 200 OK\"\n",
      "\u001b[92m15:50:28 - LiteLLM:INFO\u001b[0m: utils.py:2896 - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n",
      "2025-04-02 15:50:28,842 - LiteLLM - INFO - \n",
      "LiteLLM completion() model= deepseek-r1:14b; provider = ollama\n"
     ]
    }
   ],
   "source": [
    "!structsense-cli extract --agentconfig ner_config/ner_agent.yaml --taskconfig ner_config/ner_task.yaml --embedderconfig ner_config/embedding.yaml --flowconfig ner_config/flow_ner.yaml --knowledgeconfig ner_config/search_ontology_knowledge.yaml --source test.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60a9a29d-bcd7-4814-b3fc-404ba231ae60",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38df9a8a-a2e3-48ad-aba7-d6766a6777c3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
